promotion-guide.adoc

```diff
@@ -0,0 +1,137 @@
+
+As mentioned in the "Source Code Organisation" section, the MRO pipeline allows to enable separate development environments to isolate different version. When this mode is enabled, pipeline runs with `version=WIP` will deploy into the `$PROJECT-dev` as usual, but pipeline runs with `version=X` will deploy into `$PROJECT-dev-X`. If `$PROJECT-dev-X` does not exist yet, it will be created on the fly (by cloning `$PROJECT-dev` with its serviceaccounts and rolebindings). For the creation to work, the `jenkins` serviceaccount running the pipeline needs to have `self-provisioner` permissions (which need to be granted by a cluster admin). To enable (up to three) separate development environments, you can set `versionedDevEnvs` to `true` in the config map of your `Jenkinsfile`, like this:
+
+```
+def config = [debug: true, odsImageTag: 'x.x', versionedDevEnvs: true]
+```
+
+
+## Walkthrough
+
+Let's start by assuming you have a project FOO with two components, X and Y. These components are defined under the `repositories` section in the `metadata.yml` file of the release manager repository. When you want to create a new release, you start the MRO pipeline with input parameters - we will use version `1` and change ID `1234` in this example. The environment should be `DEV`. At the end of the pipeline run, you'll have a release bundle identified by the tag `v1-1234-0-D`. This release can later be promoted as-is to QA. Once it is installed there, the same release bundle will be tagged with `v1-1234-0-Q` which can then be promoted to PROD (where it will be tagged with `v1-1234-0-P`).
+
+To create a release bundle, the MRO will first trigger the build of each component. Then, it will export all resources in your OpenShift namespace (`$PROJECT-$ENVIRONMENT`, here `foo-dev`) belonging to the component. By convention, this means all resources labeled with `app=$PROJECT-$COMPONENT` (e.g. `app=foo-x`). Any resources without such a label will NOT be part of the release bundle. The exported resources are stored in a `template.yml` file (an OpenShift template) located in the `openshift-exported` folder. Further, the container image SHA of the running pod is retrieved and stored in the file `image-sha` in the same folder. Once done, the MRO will commit the two files, tag the commit with `v1-1234-0-D` and push to the remote. After this process has been done for all repositories, the same tag is also applied to the release manager repository. At this stage, the "dev release bundle" is complete and can be installed into QA.
```

Comment:
Do people need to do something extra to make sure the components are included in the release bundle? It seems like the variable `app=$PROJECT-$COMPONENT` is set during project creation, right?

Reply:
That is correct, that label is applied when a quickstarter is provisioned. They only need to add this label when they add more resources via the UI.